cpp/src/arrow/types/struct.cc

```diff
@@ -17,4 +17,94 @@
 
 #include "arrow/types/struct.h"
 
-namespace arrow {}  // namespace arrow
+#include <sstream>
+
+namespace arrow {
+
+bool StructArray::EqualsExact(const StructArray& other) const {
+  if (this == &other) { return true; }
+  if (null_count_ != other.null_count_) { return false; }
+
+  bool equal_null_bitmap = true;
+  if (null_count_ > 0) {
+    equal_null_bitmap =
+        null_bitmap_->Equals(*other.null_bitmap_, util::bytes_for_bits(length_));
+  }
+
+  if (!equal_null_bitmap) { return false; }
+  bool values_equal = false;
+  for (size_t i = 0; i < field_arrays_.size(); ++i) {
+    values_equal = field_arrays_.at(i).get()->Equals(other.fields().at(i));
```

Comment:
@wesm What was the intent of EqualsExact? Was it logical equality or physical equality.   If its logical equality we will need to add some additional methods to avoid memory allocations for constructing the logical & bitmaps.  If it is physical equality then this implementation is fine.